DaisyUI v5 Theme for React JSON Schema Form v6

inchoate opened the pull request:
# Integrate DaisyUI v5 with RJSF v6

This PR integrates DaisyUI v5 with React JSON Schema Form v6, providing a complete theme implementation with enhanced visual styling and component architecture.

## Changes

- Implemented all required template components for RJSF v6 compatibility 
- Added support for DaisyUI v6 utility classes across all form elements
- Created connected card styling for array items with proper visual hierarchy
- Enhanced field layout with consistent spacing and alignment
- Implemented specialized widget rendering for improved cross-browser compatibility
- Fixed checkbox handling to avoid double-labeling issues
- Optimized button positioning and styling
- Added proper handling of error states and validation messages
- Included comprehensive documentation of component architecture
- Added a few more fields to `nested.ts` to enhance the demo
- Fixes Selects
- Fixes Calendars
- Uses Fieldsets for arrays instead of custom array rendering
- Using new linter settings, so should have proper newlines everywhere
- Passes linting
- Passes cs-format

## Testing

Tested across multiple browsers with complex nested schemas including:
- Deeply nested objects
- Array fields with dynamic items
- Various input types (text, numbers, checkboxes, dates)
- Error validation states

## Screenshots

<img width="443" alt="Screenshot 2025-04-04 at 10 58 01 PM" src="https://github.com/user-attachments/assets/769d2ccd-41c5-4a94-a9e5-6de57e5747e9" />
<img width="463" alt="Screenshot 2025-04-04 at 10 57 47 PM" src="https://github.com/user-attachments/assets/110aef09-fb43-4a45-97ca-5f29893eeea5" />
<img width="449" alt="Screenshot 2025-04-04 at 10 57 20 PM" src="https://github.com/user-attachments/assets/2f3319a2-88d8-4ed2-b9c7-8017874f8ca1" />
<img width="424" alt="Screenshot 2025-04-04 at 10 56 59 PM" src="https://github.com/user-attachments/assets/fcec54ec-0db0-4f73-a1e6-9d3861f87134" />
<img width="474" alt="Screenshot 2025-04-04 at 10 55 13 PM" src="https://github.com/user-attachments/assets/a9ac6a51-9952-4985-a951-6f41c24162a3" />
<img width="481" alt="Screenshot 2025-04-04 at 10 55 01 PM" src="https://github.com/user-attachments/assets/9caca815-07d6-4155-b884-129d7e5e79fa" />
<img width="308" alt="Screenshot 2025-04-04 at 10 54 45 PM" src="https://github.com/user-attachments/assets/bea91c26-72e6-4d1b-8e11-da273350fe45" />
<img width="475" alt="Screenshot 2025-04-04 at 10 54 37 PM" src="https://github.com/user-attachments/assets/1b69bfc9-1da0-4080-952f-a5c50056725d" />

@heath-freenome @saadeghi

inchoate commented on packages/daisyui/README.md:
```diff
@@ -0,0 +1,70 @@
+
+A [DaisyUI](https://daisyui.com/) theme for [react-jsonschema-form](https://github.com/rjsf-team/react-jsonschema-form/).
+
+**Warning:** This integrates DaisyUI v5.0.0-beta.7. This is not yet released and is subject to change. It also integrates tailwindcss v4.0.6.
```
This is a lie. It's no longer beta.

TODO

heath-freenome commented on packages/playground/vite.config.ts:
```diff
@@ -26,6 +26,12 @@ export default defineConfig({
       // We want to pick up the browser version of the utils
       '@rjsf/utils': path.resolve(__dirname, '../utils/src/indexBrowser'),
       '@rjsf/validator-ajv8': path.resolve(__dirname, '../validator-ajv8/src'),
+      '@rjsf/daisyui': path.resolve(__dirname, '../daisyui/src'),
+      // validator-ajv6 can not be mapped directly to the sources, because that causes wrong ajv version resolution
+      // which looks related to: https://github.com/vitejs/vite/issues/12618
+      //
+      // the difference when mapping directly vs mapping to src folder - @vitejs/plugin-react can not be applied in the 2nd case
+      '@rjsf/validator-ajv6': '@rjsf/validator-ajv6/src',
```
We deleted this validator in v6

heath-freenome replied:
```suggestion
      '@rjsf/daisyui': path.resolve(__dirname, '../daisyui/src'),
```

heath-freenome commented on tsconfig.json:
```diff
@@ -41,4 +44,4 @@
     }
   ]
-}
+}
```
```suggestion
}

```

heath-freenome commented on packages/daisyui/src/widgets/TextAreaWidget/TextAreaWidget.tsx:
```diff
@@ -0,0 +1,21 @@
+  S extends StrictRJSFSchema = RJSFSchema,
+  F extends FormContextType = any
+>(props: WidgetProps<T, S, F>) {
+  const { id, value, required, disabled, readonly, onChange } = props;
```
Please also handle `onFocus` and `onBlur`

heath-freenome commented on packages/daisyui/src/widgets/RatingWidget/RatingWidget.tsx:
```diff
@@ -0,0 +1,42 @@
+  T = any,
+  S extends StrictRJSFSchema = RJSFSchema,
+  F extends FormContextType = any
+>({ id, value, required, disabled, readonly, autofocus, onChange, schema, options }: WidgetProps<T, S, F>) {
```
Please also handle `onFocus` and `onBlur`

heath-freenome commented on packages/daisyui/src/widgets/RangeWidget/RangeWidget.tsx:
```diff
@@ -0,0 +1,31 @@
+  required,
+  disabled,
+  readonly,
+  onChange,
```
Please also handle `onFocus` and `onBlur`

heath-freenome commented on packages/daisyui/src/widgets/DateWidget/DateWidget.tsx:
```diff
@@ -0,0 +1,220 @@
+const DateTimeWidget = <T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
+  props: WidgetProps<T, S, F>
+) => {
+  const { id, value, onChange, schema } = props;
```
Please also handle `onFocus` and `onBlur`

heath-freenome commented on packages/daisyui/src/widgets/DateTimeWidget/DateTimeWidget.tsx:
```diff
@@ -0,0 +1,250 @@
+const DateTimeWidget = <T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
+  props: WidgetProps<T, S, F>
+) => {
+  const { id, value, onChange, schema } = props;
```
Please also handle `onFocus` and `onBlur`

heath-freenome commented on packages/daisyui/src/widgets/CheckboxesWidget/CheckboxesWidget.tsx:
```diff
@@ -0,0 +1,57 @@
+            checked={isChecked(option)}
+            required={required}
+            disabled={disabled || readonly}
+            onChange={() => _onChange(option)}
```
This pattern causes a rerender every time. Consider using `useCallback()`. Please also handle `onFocus` and `onBlur`.

heath-freenome commented on packages/daisyui/src/widgets/CheckboxWidget/CheckboxWidget.tsx:
```diff
@@ -0,0 +1,22 @@
+  S extends StrictRJSFSchema = RJSFSchema,
+  F extends FormContextType = any
+>(props: WidgetProps<T, S, F>) {
+  const { id, value, required, disabled, readonly, onChange } = props;
```
Please also handle `onFocus` and `onBlur`

heath-freenome commented on packages/daisyui/src/templates/Templates.tsx:
```diff
@@ -0,0 +1,49 @@
+    ErrorListTemplate: ErrorList,
+    FieldErrorTemplate,
+    FieldHelpTemplate,
+    FieldTemplate,
```
You are missing the new `GridTemplate` implementation. Be sure to update the `layoutGrid.ts` file in the `playground` once you provide it.

inchoate commented:
- Addressing PR feedback
- Added Rating widget to the demo
<img width="678" alt="Screenshot 2025-04-08 at 9 32 22 AM" src="https://github.com/user-attachments/assets/038de634-4dc4-4f04-86f0-6d42d3bacf94" />

- TODO: GridTemplate

inchoate commented:
- Added GridLayout
- Fixed Calendar Rendering
- Addressed all PR issues

heath-freenome commented on packages/daisyui/babel.config.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "extends": "../../babel.config.json"
+}
```
Adding blank line at end to remove the red circled dash
```suggestion
}

```

inchoate replied:
Fighting my linter. I'll fix.

heath-freenome commented on packages/daisyui/package.json:
```diff
@@ -0,0 +1,80 @@
+    "tailwindcss": "^4.0.6"
+  },
+  "peerDependencies": {
+    "react": "^18.2.0",
```
All of the other packages have react dependencies set to `>=17`. Does this one require `^18.2`?

heath-freenome commented on packages/daisyui/package.json:
```diff
@@ -0,0 +1,80 @@
+    "primereact": ">=8.0.0"
+  },
+  "devDependencies": {
+    "daisyui": "^5.0.0-beta.7",
```
I thought you mentioned it was no longer a beta?

inchoate replied:
5.0.17 is the latest and is non-beta. Will fix.

heath-freenome commented on packages/daisyui/package.json:
```diff
@@ -0,0 +1,80 @@
+    "type": "git",
+    "url": "git+https://github.com/rjsf-team/react-jsonschema-form.git"
+  },
+  "author": "Your Name <your.email@example.com>",
```
Did you want to add your contact info?

heath-freenome commented on packages/daisyui/src/daisyForm/index.ts:
```diff
@@ -0,0 +1,2 @@
+export { default } from './DaisyForm';
+export * from './DaisyForm';
```
Adding blank line at end to remove the red circled dash
```suggestion
export * from './DaisyForm';

```

heath-freenome commented on packages/daisyui/src/styles.css:
```diff
@@ -0,0 +1,9 @@
+    --color-primary: 255 115 179;
+    --color-secondary: 111 114 185;
+  }
+} 
```
Adding blank line at end to remove the red circled dash
```suggestion
} 

```

heath-freenome commented on packages/daisyui/src/templates/GridTemplate/index.ts:
```diff
@@ -0,0 +1 @@
+export { default } from './GridTemplate'; 
```
Adding blank line at end to remove the red circled dash
```suggestion
export { default } from './GridTemplate'; 
export * from './GridTemplate';

```

heath-freenome commented on packages/daisyui/src/tsconfig.json:
```diff
@@ -0,0 +1,26 @@
+      "path": "../../validator-ajv8"
+    }
+  ]
+}
```
Adding blank line at end to remove the red circled dash
```suggestion
}

```

heath-freenome commented on packages/daisyui/src/types.d.ts:
```diff
@@ -0,0 +1,4 @@
+declare module '*.css?inline' {
+    const content: string;
```
Hmmm, I'm surprised the linter didn't change the indents to 2 spaces

inchoate replied:
Agreed. It's forcing to 4-spaces here. Will make two.

heath-freenome commented on packages/daisyui/src/types/DaisyProps.ts:
```diff
@@ -0,0 +1,5 @@
+export interface DaisyProps {
+    theme?: string;
```
Hmmm, I'm surprised the linter didn't change the indents to 2 spaces

heath-freenome commented on packages/daisyui/src/utils.ts:
```diff
@@ -0,0 +1,17 @@
+import { UiSchema } from '@rjsf/utils';
+
+export interface DaisyUiSchema extends Omit<UiSchema, 'ui:options'> {
+    'ui:options'?: DaisyUiOptions;
```
Hmmm, I'm surprised the linter didn't change the indents to 2 spaces

heath-freenome commented on packages/daisyui/src/widgets/TextAreaWidget/index.ts:
```diff
@@ -0,0 +1,4 @@
+import TextareaWidget from './TextAreaWidget';
+
+export default TextareaWidget;
+
```
```suggestion
export default TextareaWidget;
```

heath-freenome commented on packages/daisyui/src/widgets/ToggleWidget/index.ts:
```diff
@@ -0,0 +1,2 @@
+export { default } from './ToggleWidget';
+export * from './ToggleWidget';
```
Adding blank line at end to remove the red circled dash
```suggestion
export * from './ToggleWidget';

```

heath-freenome commented on packages/daisyui/tailwind.config.js:
```diff
@@ -0,0 +1,7 @@
+// Add plugins that require JavaScript functions
+config.plugins = [require("daisyui")];
+
+module.exports = config; 
```
Adding blank line at end to remove the red circled dash
```suggestion
module.exports = config; 

```

heath-freenome commented on packages/daisyui/tsconfig.build.json:
```diff
@@ -0,0 +1,31 @@
+        "lib",
+        "dist"
+    ]
+}
```
Adding blank line at end to remove the red circled dash
```suggestion
}

```

heath-freenome commented on packages/daisyui/tsconfig.json:
```diff
@@ -0,0 +1,24 @@
+            "path": "../validator-ajv8"
+        }
+    ]
+}
```
Adding blank line at end to remove the red circled dash
```suggestion
}

```

heath-freenome commented on packages/playground/src/samples/widgets.tsx:
```diff
@@ -131,6 +193,48 @@ const widgets: Sample = {
     },
+    customWidgets: {
+      rating: {
+        'ui:widget': 'rating',
```
What if the theme doesn't have a `rating` widget. Does it still work properly?

heath-freenome commented on packages/playground/tsconfig.json:
```diff
@@ -19,18 +28,45 @@
+    }
   ]
-}
+}
```
Adding blank line at end to remove the red circled dash
```suggestion
}

```

inchoate commented on packages/daisyui/src/templates/Templates.tsx:
```diff
@@ -0,0 +1,66 @@
+
+import { FormContextType, RJSFSchema, StrictRJSFSchema, TemplatesType } from '@rjsf/utils';
+
+import AddButton from './ButtonTemplates/AddButton';
+import SubmitButton from './ButtonTemplates/SubmitButton';
+import { CopyButton, MoveDownButton, MoveUpButton, RemoveButton } from './ButtonTemplates/IconButton';
```
Yes, absolutely.

inchoate commented:
@heath-freenome I think we're closer now.

- Fixed almost all linter issues. Some I'm okay w/.
- (feat) / Added new RatingWidget across the board and added it to the 'Widgets' demo
- Addressed all of your concerns. Please note, however, that my IDE linter is overriding yours so a newline or two might have been eaten somew
here. Happy to fix whatever snuck through.
- This also fixed up the calendar some more.


# Core Rating Widget
<img width="442" alt="Screenshot 2025-04-09 at 6 25 08 AM" src="https://github.com/user-attachments/assets/a29ea682-24fc-4194-b9a7-353d615273a0" />


# Default Style
<img width="506" alt="Screenshot 2025-04-09 at 6 24 22 AM" src="https://github.com/user-attachments/assets/9f9bb13c-5f4e-48c4-bee6-5b881327a18d" />

# Daisy Style ❤️ 
<img width="401" alt="Screenshot 2025-04-09 at 6 25 35 AM" src="https://github.com/user-attachments/assets/f53afb82-c5a4-4032-add2-2586e3bfb967" />

heath-freenome commented on packages/daisyui/src/templates/FieldTemplate/FieldTemplate.tsx:
```diff
@@ -0,0 +1,77 @@
+ *
+ * @param props - The `FieldTemplateProps` for the component
+ */
+function FieldTemplate<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
+  props: FieldTemplateProps<T, S, F>
+) {
```
```suggestion
export default function FieldTemplate<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: FieldTemplateProps<T, S, F>
) {
```

heath-freenome replied:
@inchoate this one slipped past

heath-freenome commented on packages/daisyui/src/templates/FieldTemplate/FieldTemplate.tsx:
```diff
@@ -0,0 +1,77 @@
+      {help}
+    </div>
+  );
+}
+
+export default FieldTemplate;
```
```suggestion
}
```

heath-freenome replied:
@inchoate missed this one too

heath-freenome commented on packages/daisyui/src/templates/Templates.tsx:
```diff
@@ -0,0 +1,64 @@
+import ArrayFieldItemTemplate from './ArrayFieldItemTemplate';
+import ArrayFieldTemplate from './ArrayFieldTemplate/ArrayFieldTemplate';
+import ArrayFieldItemButtonsTemplate from './ArrayFieldItemButtonsTemplate';
+import BaseInputTemplate from './BaseInputTemplate/BaseInputTemplate';
+import DescriptionField from './DescriptionField';
+import ErrorList from './ErrorList';
+import FieldErrorTemplate from './FieldErrorTemplate';
+import FieldHelpTemplate from './FieldHelpTemplate';
+import FieldTemplate from './FieldTemplate';
+import GridTemplate from './GridTemplate/GridTemplate';
+import ObjectFieldTemplate from './ObjectFieldTemplate';
+import TitleFieldTemplate from './TitleField/TitleField';
+import WrapIfAdditionalTemplate from './WrapIfAdditionalTemplate';
+
+import { FormContextType, RJSFSchema, StrictRJSFSchema, TemplatesType } from '@rjsf/utils';
+
+import { AddButton, SubmitButton, CopyButton, MoveDownButton, MoveUpButton, RemoveButton } from './ButtonTemplates';
```
```suggestion
import { FormContextType, RJSFSchema, StrictRJSFSchema, TemplatesType } from '@rjsf/utils';

import ArrayFieldItemTemplate from './ArrayFieldItemTemplate';
import ArrayFieldTemplate from './ArrayFieldTemplate/ArrayFieldTemplate';
import ArrayFieldItemButtonsTemplate from './ArrayFieldItemButtonsTemplate';
import BaseInputTemplate from './BaseInputTemplate/BaseInputTemplate';
import DescriptionField from './DescriptionField';
import ErrorList from './ErrorList';
import FieldErrorTemplate from './FieldErrorTemplate';
import FieldHelpTemplate from './FieldHelpTemplate';
import FieldTemplate from './FieldTemplate';
import GridTemplate from './GridTemplate/GridTemplate';
import ObjectFieldTemplate from './ObjectFieldTemplate';
import TitleFieldTemplate from './TitleField/TitleField';
import WrapIfAdditionalTemplate from './WrapIfAdditionalTemplate';
import { AddButton, SubmitButton, CopyButton, MoveDownButton, MoveUpButton, RemoveButton } from './ButtonTemplates';
```

heath-freenome commented on packages/daisyui/src/templates/WrapIfAdditionalTemplate/WrapIfAdditionalTemplate.tsx:
```diff
@@ -0,0 +1,62 @@
+ *
+ * @param props - The `WrapIfAdditionalProps` for this component
+ */
+function WrapIfAdditionalTemplate<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
+  props: WrapIfAdditionalTemplateProps<T, S, F>
+) {
```
```suggestion
export default function WrapIfAdditionalTemplate<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: WrapIfAdditionalTemplateProps<T, S, F>
) {
```

heath-freenome commented on packages/daisyui/src/templates/WrapIfAdditionalTemplate/WrapIfAdditionalTemplate.tsx:
```diff
@@ -0,0 +1,62 @@
+      {children}
+    </div>
+  );
+}
+
+export default WrapIfAdditionalTemplate;
```
```suggestion
}
```

heath-freenome commented on packages/daisyui/src/widgets/DateTimeWidget/DateTimeWidget.tsx:
```diff
@@ -0,0 +1,326 @@
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { format, isValid, isToday, isSameDay } from 'date-fns';
+import { DayPicker, ClassNames, ModifiersClassNames, UI } from 'react-day-picker';
+import 'react-day-picker/dist/style.css';
+
+import { FormContextType, RJSFSchema, StrictRJSFSchema, WidgetProps } from '@rjsf/utils';
```
```suggestion
import 'react-day-picker/dist/style.css';
import { FormContextType, RJSFSchema, StrictRJSFSchema, WidgetProps } from '@rjsf/utils';
```

heath-freenome commented on packages/daisyui/src/widgets/DateWidget/DateWidget.tsx:
```diff
@@ -0,0 +1,289 @@
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { format, isValid, isToday, isSameDay } from 'date-fns';
+import { DayPicker, ClassNames, ModifiersClassNames, UI } from 'react-day-picker';
+import 'react-day-picker/dist/style.css';
+
+import { FormContextType, RJSFSchema, StrictRJSFSchema, WidgetProps } from '@rjsf/utils';
```
```suggestion
import 'react-day-picker/dist/style.css';
import { FormContextType, RJSFSchema, StrictRJSFSchema, WidgetProps } from '@rjsf/utils';
```

heath-freenome commented on packages/daisyui/src/widgets/RatingWidget/index.ts:
```diff
@@ -0,0 +1,3 @@
+import RatingWidget from './RatingWidget';
+
+export default RatingWidget;
```
```suggestion
export { default } from './RatingWidget';
export * from './RatingWidget';
```

heath-freenome commented on packages/daisyui/src/widgets/SelectWidget/SelectWidget.tsx:
```diff
@@ -0,0 +1,121 @@
+} from '@rjsf/utils';
+import { ChangeEvent, FocusEvent } from 'react';
+
+export default function SelectWidget<
```
@inchoate Please complete this

heath-freenome commented on packages/daisyui/src/widgets/Widgets.tsx:
```diff
@@ -0,0 +1,57 @@
+import { FormContextType, RegistryWidgetsType, RJSFSchema, StrictRJSFSchema } from '@rjsf/utils';
+import AltDateTimeWidget from './AltDateTimeWidget/AltDateTimeWidget';
```
```suggestion
import { FormContextType, RegistryWidgetsType, RJSFSchema, StrictRJSFSchema } from '@rjsf/utils';

import AltDateTimeWidget from './AltDateTimeWidget/AltDateTimeWidget';
```

heath-freenome commented on packages/daisyui/tsconfig.build.json:
```diff
@@ -0,0 +1,32 @@
+        "lib",
+        "dist"
+    ]
+}
+
```
```suggestion
}
```

heath-freenome commented on packages/playground/src/samples/examples.ts:
```diff
@@ -1,6 +1,8 @@
 import { Sample } from './Sample';
+import validator from '@rjsf/validator-ajv8';
```
```suggestion
import validator from '@rjsf/validator-ajv8';

import { Sample } from './Sample';
```

inchoate replied:
@heath-freenome how was this passing the linter before if `validator` was imported but not being used?

heath-freenome replied:
?! good question Thanks for fixing

heath-freenome commented on packages/core/src/index.ts:
```diff
@@ -1,8 +1,9 @@
 import withTheme, { ThemeProps } from './withTheme';
 import getDefaultRegistry from './getDefaultRegistry';
+import RatingWidget from './components/widgets/RatingWidget';
 
 export type { FormProps, FormState, IChangeEvent, ThemeProps };
 
-export { withTheme, getDefaultRegistry };
+export { withTheme, getDefaultRegistry, RatingWidget };
```
You can revert this change

inchoate replied:
Good catch.

heath-freenome commented on packages/daisyui/src/templates/ArrayFieldTemplate/ArrayFieldTemplate.tsx:
```diff
@@ -0,0 +1,108 @@
+          <div className='flex justify-end'>
+            <AddButton
+              className='btn btn-primary btn-sm'
+              onClick={(e) => {
+                e.preventDefault();
+                e.stopPropagation();
+                onAddClick(e);
+              }}
```
Consider moving complex callbacks into an external function, perhaps using `useCallback()` to avoid rerenders

heath-freenome commented on packages/daisyui/src/templates/FieldHelpTemplate/FieldHelpTemplate.tsx:
```diff
@@ -0,0 +1,23 @@
+  const { help } = props;
+  return (
+    <div className='field-help-template text-gray-500 text-sm'>
+      <p>{help}</p>
```
Something to consider, if `help` contains something other than text (i.e. something with other HTML tags), it will cause DOM warnings about other html elements inside of a `<p>`. I've had to update code to detect whether `help` is something other than a string and switch from `<p>` to `<div>`. Or is it ok to just use `<div>` here outright?

heath-freenome commented on packages/daisyui/src/templates/DescriptionField/DescriptionField.tsx:
```diff
@@ -0,0 +1,20 @@
+  const { id, description } = props;
+  return (
+    <div id={id} className='description-field my-4'>
+      <p className='text-sm text-base-content/80'>{description}</p>
```
Something to consider, if `description` contains something other than text (i.e. something with other HTML tags), it will cause DOM warnings about other html elements inside of a `<p>`. I've had to update code to detect whether `description` is something other than a string and switch from `<p>` to `<div>`. Or is it ok to just use `<div>` here outright?

heath-freenome commented on packages/daisyui/src/widgets/CheckboxWidget/CheckboxWidget.tsx:
```diff
@@ -0,0 +1,56 @@
+      checked={value}
+      required={required}
+      disabled={disabled || readonly}
+      onChange={(event) => onChange(event.target.checked)}
```
Consider making a wrapper function for this one too?

heath-freenome commented on packages/daisyui/src/widgets/DateTimeWidget/DateTimeWidget.tsx:
```diff
@@ -0,0 +1,325 @@
+              <button
+                type='button'
+                className='btn btn-sm btn-primary'
+                onClick={() => {
+                  setIsOpen(false);
+                  onChange(localDate ? localDate.toISOString() : '');
+                  if (onBlur) {
+                    onBlur(id, value);
+                  }
+                  inputRef.current?.focus();
+                }}
```
Consider moving complex callbacks into an external function, perhaps using `useCallback()` to avoid rerenders

heath-freenome commented on packages/daisyui/src/widgets/DateWidget/DateWidget.tsx:
```diff
@@ -0,0 +1,391 @@
+              <button
+                type='button'
+                className='btn btn-sm btn-primary'
+                onClick={() => {
+                  setIsOpen(false);
+                  onChange(localDate ? localDate.toISOString() : '');
+                  if (onBlur) {
+                    onBlur(id, value);
+                  }
+                  inputRef.current?.focus();
+                }}
```
Consider moving complex callbacks into an external function, perhaps using `useCallback()` to avoid rerenders

heath-freenome commented on packages/daisyui/src/widgets/RadioWidget/RadioWidget.tsx:
```diff
@@ -0,0 +1,102 @@
+              checked={isChecked(option)}
+              required={required}
+              disabled={disabled || readonly}
+              onChange={(e) => {
+                onChange(isEnumeratedObject ? option.value : option.value);
+                e.target.blur();
+              }}
```
Consider moving complex callbacks into an external function, perhaps using `useCallback()` to avoid rerenders

heath-freenome commented on packages/daisyui/src/widgets/RangeWidget/RangeWidget.tsx:
```diff
@@ -0,0 +1,67 @@
+          min={schema.minimum}
+          max={schema.maximum}
+          step={schema.multipleOf || 1}
+          onChange={(event) => onChange(event.target.value)}
```
Consider adding a wrapper function like the others?

heath-freenome commented on packages/daisyui/src/widgets/TextAreaWidget/TextAreaWidget.tsx:
```diff
@@ -0,0 +1,60 @@
+        disabled={disabled || readonly}
+        readOnly={readonly}
+        rows={rows}
+        onChange={(event) => onChange(event.target.value)}
```
consider a wrapper function like the others?

heath-freenome commented on packages/daisyui/src/widgets/TimeWidget/TimeWidget.tsx:
```diff
@@ -0,0 +1,55 @@
+        required={required}
+        disabled={disabled || readonly}
+        readOnly={readonly}
+        onChange={(event) => onChange(event.target.value)}
```
Consider a wrapper function like the others?

heath-freenome commented on packages/daisyui/tailwind.config.js:
```diff
@@ -0,0 +1,20 @@
+/** @type {import('tailwindcss').Config} */
```
So I'm guessing this is working without the conversion

heath-freenome commented on packages/daisyui/package.json:
```diff
@@ -0,0 +1,54 @@
+  "engines": {
+    "node": ">=20"
+  },
+  "scripts": {
+    "build": "tsc -b tsconfig.build.json",
+    "dev": "vite"
```
These scripts are all missing. Note you will need to copy the `tsconfig.build.json` file from `chakra-ui`:
```
"build:ts": "tsc -b tsconfig.build.json && tsc-alias -p tsconfig.build.json",
    "build:cjs": "esbuild ./src/index.ts --bundle --outfile=dist/index.js --sourcemap --packages=external --format=cjs",
    "build:esm": "esbuild ./src/index.ts --bundle --outfile=dist/chakra-ui.esm.js --sourcemap --packages=external --format=esm",
    "build:umd": "rollup dist/chakra-ui.esm.js --format=umd --file=dist/chakra-ui.umd.js --name=@rjsf/chakra-ui",
    "build": "npm run build:ts && npm run build:cjs && npm run build:esm && npm run build:umd",
    "cs-check": "prettier -l \"{src,test}/**/*.ts?(x)\"",
    "cs-format": "prettier \"{src,test}/**/*.ts?(x)\" --write",
    "lint": "eslint src test",
    "precommit": "lint-staged",
    "test": "jest",
    "test:update": "jest --u",
```

heath-freenome commented on packages/daisyui/src/widgets/DateTimeWidget/DateTimeWidget.tsx:
```diff
@@ -0,0 +1,326 @@
+      <div
+        className='w-full'
+        tabIndex={0}
+        onKeyDown={(e) => {
+          if (e.key === 'Enter' || e.key === ' ') {
+            togglePicker(e as unknown as React.MouseEvent);
+          }
+        }}
```
Another `useCallback()` candidate

heath-freenome commented on packages/daisyui/src/widgets/DateWidget/DateWidget.tsx:
```diff
@@ -0,0 +1,392 @@
+      <div
+        className='w-full'
+        tabIndex={0}
+        onKeyDown={(e) => {
+          if (e.key === 'Enter' || e.key === ' ') {
+            togglePicker(e as unknown as React.MouseEvent);
+          }
+        }}
```
Another `useCallback()` candidate

heath-freenome commented on packages/daisyui/src/widgets/FileWidget/FileWidget.tsx:
```diff
@@ -0,0 +1,80 @@
+   *
+   * @param event - The blur event
+   */
+  const handleBlur = (event: FocusEvent<HTMLInputElement>) => {
+    if (onBlur) {
+      onBlur(id, event.target.files ? Array.from(event.target.files) : null);
+    }
+  };
```
Another `useCallback()` candidate

heath-freenome commented on packages/daisyui/src/widgets/FileWidget/FileWidget.tsx:
```diff
@@ -0,0 +1,80 @@
+   *
+   * @param event - The focus event
+   */
+  const handleFocus = (event: FocusEvent<HTMLInputElement>) => {
+    if (onFocus) {
+      onFocus(id, event.target.files ? Array.from(event.target.files) : null);
+    }
+  };
```
Another `useCallback()` candidate

heath-freenome commented on packages/daisyui/src/widgets/RadioWidget/RadioWidget.tsx:
```diff
@@ -0,0 +1,112 @@
+   * @param event - The blur event
+   * @param optionValue - The value of the blurred option
+   */
+  const handleBlur = (event: FocusEvent<HTMLInputElement>, optionValue: any) => {
+    if (onBlur) {
+      onBlur(id, optionValue);
+    }
+  };
```
Another `useCallback()` candidate

heath-freenome commented on packages/daisyui/src/theme/Theme.tsx:
```diff
@@ -0,0 +1,93 @@
+ *
+ * @returns The current theme context with theme name and setter function
+ */
+export const useTheme = () => React.useContext(ThemeContext);
```
```suggestion
export const useTheme = () => useContext(ThemeContext);
```

heath-freenome commented on packages/daisyui/src/widgets/CheckboxesWidget/CheckboxesWidget.tsx:
```diff
@@ -0,0 +1,124 @@
+import { WidgetProps, StrictRJSFSchema, RJSFSchema, FormContextType } from '@rjsf/utils';
+import { FocusEvent, useCallback } from 'react';
```
```suggestion
import { FocusEvent, useCallback } from 'react';
import { WidgetProps, StrictRJSFSchema, RJSFSchema, FormContextType } from '@rjsf/utils';
```

heath-freenome commented on packages/daisyui/src/widgets/DateTimeWidget/DateTimeWidget.tsx:
```diff
@@ -0,0 +1,342 @@
+import { faCalendar } from '@fortawesome/free-solid-svg-icons';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { FormContextType, RJSFSchema, StrictRJSFSchema, WidgetProps } from '@rjsf/utils';
+import { format, isSameDay, isToday, isValid } from 'date-fns';
+import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
```
```suggestion
import { RefObject, useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { faCalendar } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { FormContextType, RJSFSchema, StrictRJSFSchema, WidgetProps } from '@rjsf/utils';
import { format, isSameDay, isToday, isValid } from 'date-fns';
```

heath-freenome commented on packages/daisyui/src/widgets/DateTimeWidget/DateTimeWidget.tsx:
```diff
@@ -0,0 +1,342 @@
+ * @param ref - React ref to the element to monitor
+ * @param callback - Function to call when a click outside is detected
+ */
+function useClickOutside(ref: React.RefObject<HTMLDivElement>, callback: () => void) {
```
```suggestion
function useClickOutside(ref: RefObject<HTMLDivElement>, callback: () => void) {
```

heath-freenome commented on packages/daisyui/src/widgets/DateWidget/DateWidget.tsx:
```diff
@@ -0,0 +1,393 @@
+import { faCalendar } from '@fortawesome/free-solid-svg-icons';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { FormContextType, RJSFSchema, StrictRJSFSchema, WidgetProps } from '@rjsf/utils';
+import { format, isSameDay, isToday, isValid } from 'date-fns';
+import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
```
```suggestion
import { RefObject, useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { faCalendar } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { FormContextType, RJSFSchema, StrictRJSFSchema, WidgetProps } from '@rjsf/utils';
import { format, isSameDay, isToday, isValid } from 'date-fns';
```

heath-freenome commented on packages/daisyui/src/widgets/FileWidget/FileWidget.tsx:
```diff
@@ -0,0 +1,86 @@
+   * @param event - The change event from the file input
+   */
+  const handleChange = useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
```
```suggestion
    (event: ChangeEvent<HTMLInputElement>) => {
```

heath-freenome commented on packages/daisyui/src/widgets/RadioWidget/RadioWidget.tsx:
```diff
@@ -0,0 +1,114 @@
+import { WidgetProps, StrictRJSFSchema, FormContextType, RJSFSchema } from '@rjsf/utils';
+import { FocusEvent, useCallback } from 'react';
```
```suggestion
import { FocusEvent, useCallback } from 'react';
import { WidgetProps, StrictRJSFSchema, FormContextType, RJSFSchema } from '@rjsf/utils';
```

heath-freenome commented on packages/daisyui/src/widgets/RangeWidget/RangeWidget.tsx:
```diff
@@ -0,0 +1,78 @@
+import { WidgetProps, StrictRJSFSchema, RJSFSchema, FormContextType } from '@rjsf/utils';
+import { FocusEvent, useCallback } from 'react';
```
```suggestion
import { FocusEvent, useCallback } from 'react';
import { WidgetProps, StrictRJSFSchema, RJSFSchema, FormContextType } from '@rjsf/utils';
```

heath-freenome commented on packages/daisyui/src/widgets/SelectWidget/SelectWidget.tsx:
```diff
@@ -0,0 +1,161 @@
+import {
+  enumOptionsIndexForValue,
+  enumOptionsValueForIndex,
+  FormContextType,
+  RJSFSchema,
+  StrictRJSFSchema,
+  WidgetProps,
+} from '@rjsf/utils';
+import { FocusEvent, useCallback } from 'react';
```
```suggestion
import { FocusEvent, useCallback } from 'react';
import {
  enumOptionsIndexForValue,
  enumOptionsValueForIndex,
  FormContextType,
  RJSFSchema,
  StrictRJSFSchema,
  WidgetProps,
} from '@rjsf/utils';
```

heath-freenome commented on packages/daisyui/src/widgets/TextareaWidget/TextareaWidget.tsx:
```diff
@@ -0,0 +1,77 @@
+import { WidgetProps, StrictRJSFSchema, RJSFSchema, FormContextType } from '@rjsf/utils';
+import { FocusEvent, useCallback } from 'react';
```
```suggestion
import { FocusEvent, useCallback } from 'react';
import { WidgetProps, StrictRJSFSchema, RJSFSchema, FormContextType } from '@rjsf/utils';
```

heath-freenome commented on packages/daisyui/src/widgets/TimeWidget/TimeWidget.tsx:
```diff
@@ -0,0 +1,72 @@
+import { WidgetProps, StrictRJSFSchema, RJSFSchema, FormContextType } from '@rjsf/utils';
+import { FocusEvent, useCallback } from 'react';
```
```suggestion
import { FocusEvent, useCallback } from 'react';
import { WidgetProps, StrictRJSFSchema, RJSFSchema, FormContextType } from '@rjsf/utils';
```

inchoate commented on packages/daisyui/src/templates/ArrayFieldTemplate/ArrayFieldTemplate.tsx:
```diff
@@ -0,0 +1,118 @@
+        </div>
+        {canAdd && (
+          <div className='flex justify-end'>
+            <AddButton
+              className='btn btn-primary btn-sm'
```
Me, too.

inchoate commented:
- Addressed all PR comments
- Pinned ajv8 version to docusuarus would build cleanly
- Everything runs and builds here. Hopefully CI agrees.

```sh
$ npm run lint && npm run cs-format && nx run-many --target=build --parallel=1 --disableNxCache --skipNxCache

> lint
> eslint './packages/**/*.{ts,tsx}' --quiet


> cs-format
> nx run-many --target=cs-format


Nx is configured to use the legacy cache. This cache will be removed in Nx 21. Read more at https://nx.dev/deprecated/legacy-cache.
   ✔  nx run @rjsf/utils:cs-format  [existing outputs match the cache, left as is]
   ✔  nx run @rjsf/snapshot-tests:cs-format  [existing outputs match the cache, left as is]
   ✔  nx run @rjsf/core:cs-format  [existing outputs match the cache, left as is]
   ✔  nx run @rjsf/validator-ajv8:cs-format  [existing outputs match the cache, left as is]
   ✔  nx run @rjsf/semantic-ui:cs-format  [existing outputs match the cache, left as is]
   ✔  nx run @rjsf/chakra-ui:cs-format  [existing outputs match the cache, left as is]
   ✔  nx run @rjsf/mui:cs-format  [existing outputs match the cache, left as is]
   ✔  nx run @rjsf/fluentui-rc:cs-format  [existing outputs match the cache, left as is]
   ✔  nx run @rjsf/react-bootstrap:cs-format  [existing outputs match the cache, left as is]
   ✔  nx run @rjsf/antd:cs-format  [existing outputs match the cache, left as is]
   ✔  nx run @rjsf/playground:cs-format  [existing outputs match the cache, left as is]
   ✔  nx run @rjsf/daisyui:cs-format  [existing outputs match the cache, left as is]
   ✔  nx run @rjsf/shadcn:cs-format  [existing outputs match the cache, left as is]
   ✔  nx run @rjsf/docs:cs-format  [existing outputs match the cache, left as is]

———————————————————————————————————————————————————————————————————————————————————————————

 NX   Successfully ran target cs-format for 14 projects (68ms)

Nx read the output from the cache instead of running the command for 14 out of 14 tasks.

Nx is configured to use the legacy cache. This cache will be removed in Nx 21. Read more at https://nx.dev/deprecated/legacy-cache.
```

heath-freenome commented on packages/daisyui/src/utils.ts:
```diff
@@ -0,0 +1,17 @@
+import { DaisyProps } from './types/DaisyProps';
+import { UiSchema } from '@rjsf/utils';
+
```
```suggestion
import { UiSchema } from '@rjsf/utils';

import { DaisyProps } from './types/DaisyProps';

```

heath-freenome commented on packages/daisyui/test/DaisyUIFrameProvider.test.tsx:
```diff
@@ -0,0 +1,96 @@
+import { render } from '@testing-library/react';
+import { __createDaisyUIFrameProvider } from '../src/DaisyUIFrameProvider';
```
```suggestion
import { render } from '@testing-library/react';

import { __createDaisyUIFrameProvider } from '../src/DaisyUIFrameProvider';
```

heath-freenome commented on packages/daisyui/test/RatingWidget.test.tsx:
```diff
@@ -0,0 +1,117 @@
+import renderer from 'react-test-renderer';
+import { render, fireEvent } from '@testing-library/react';
+import RatingWidget from '../src/widgets/RatingWidget/RatingWidget';
```
```suggestion
import { render, fireEvent } from '@testing-library/react';

import RatingWidget from '../src/widgets/RatingWidget/RatingWidget';
```

heath-freenome commented on packages/daisyui/test/ToggleWidget.test.tsx:
```diff
@@ -0,0 +1,67 @@
+import renderer from 'react-test-renderer';
+import { render, fireEvent } from '@testing-library/react';
+import ToggleWidget from '../src/widgets/ToggleWidget/ToggleWidget';
```
```suggestion
import { render, fireEvent } from '@testing-library/react';

import ToggleWidget from '../src/widgets/ToggleWidget/ToggleWidget';
```
